ttracer/tracer.go

```diff
@@ -0,0 +1,95 @@
+package ttracer
+
+import (
+	"time"
+
+	ot "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+)
+
+// NewWithOptions creates a customized Tracer.
+func NewWithOptions(opts Options) ot.Tracer {
+	rval := &tracerImpl{options: opts}
+	return rval
+}
+
+// New creates and returns a standard Tracer which defers completed Spans to
+// `recorder`.
+func New(recorder SpanRecorder) ot.Tracer {
+	opts := DefaultOptions()
+	opts.Recorder = recorder
+	return NewWithOptions(opts)
+}
+
+type tracerImpl struct {
+	options Options
+}
+
+// StartSpan starts a new span with options and returns it.
+func (t *tracerImpl) StartSpan(operationName string, opts ...ot.StartSpanOption) ot.Span {
+	sso := ot.StartSpanOptions{}
+	for _, o := range opts {
+		o.Apply(&sso)
+	}
+
+	return t.StartSpanWithOptions(operationName, sso)
+}
+
+func (t *tracerImpl) StartSpanWithOptions(operationName string, opts ot.StartSpanOptions) ot.Span {
+	startTime := opts.StartTime
+	if startTime.IsZero() {
+		startTime = time.Now()
+	}
+	tags := opts.Tags
+	if tags == nil {
+		tags = map[string]interface{}{}
+	}
+
+	newSpan := t.getSpan()
+	for _, ref := range opts.References {
+		if ref.Type == ot.ChildOfRef {
+			refSpanCtx := ref.ReferencedContext.(SpanContext)
+			newSpan.raw.Context.TraceID = refSpanCtx.TraceID
+			newSpan.raw.Context.SpanID = uuid.NewV4().String()
+			newSpan.raw.ParentSpanID = refSpanCtx.SpanID
+
+			if l := len(refSpanCtx.Baggage); l > 0 {
+				newSpan.raw.Context.Baggage = make(map[string]string, l)
+				for k, v := range refSpanCtx.Baggage {
+					newSpan.raw.Context.Baggage[k] = v
+				}
+			}
+		}
+	}
+
+	if newSpan.raw.Context.TraceID == "" {
+		// Couldn't find a parent span then create a new trace and span id
+		newSpan.raw.Context.TraceID = uuid.NewV4().String()
```

Comment:
`TraceID` should not be generated here in any case. If there is no parent span, in should be able to passed in the `opts`

Reply:
If there is no parent span, `TransactionID` should be able to passed in the `opts`